Make error summary link to fields (for contact details and future forms)

kjdchapman opened the pull request:
For context, it's these links (previously hardcoded to `personal_details`):

![image](https://user-images.githubusercontent.com/429326/61884047-679ae080-aef3-11e9-8a9e-13f6c94f1c2b.png)
____
Happy to hear thoughts on whether this model name magic is _too_ magic.

duncanjbrown commented on spec/system/candidate_entering_contact_details_spec.rb:
```diff
@@ -35,9 +35,10 @@
     end
 
-    it 'sees an error summary' do
+    it 'sees an error summary with clickable links', js: true do
```
I guess we hope to avoid repeating this test for each screen! In this case it drives out the generalisation so it's useful — what's the right amount of coverage once this branch is merged, do you think?

kjdchapman replied:
Good question! I quite like testing it at this level rather than just testing the error summary component in isolation, but repeating it in every spec could be tedious. 

I've got mixed feelings on the rspec abstractions - `behaves like` and similar - for this level of spec, but that's another option.

One to think about as we go, I think?

kjdchapman commented on app/views/contact_details/_form.html.erb:
```diff
@@ -10,7 +10,7 @@
 
       <%= f.govuk_text_field :phone_number, label: { text: t('application_form.contact_details_section.phone_number.label')}, width: 10, type: :tel %>
-      <%= f.govuk_text_field :email_address, label: { text: t('application_form.contact_details_section.email_address.label')}, type: :email %>
+      <%= f.govuk_text_field :email_address, label: { text: t("application_form.contact_details_section.email_address.label")}, type: :email %>
```
Oo, my bad. I did this to test if erb-lint was enforcing quotes. It is not!

That'll teach me to `add .` for a "simple change".

kjdchapman replied:
Addressed in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/112/commits/c37a0188a5b3451404fff64380fd0d20132e5353

duncanjbrown replied:
It should, but https://github.com/Shopify/erb-lint/issues/130

kjdchapman replied:
Oh, troublesome. We're just waiting for it to get fixed upstream then?

Whilst on the subject of `erb` - I think @alexdesi was keen on a more lightweight template engine. If I remember right, you were against? Let's have a brief chat about it next week?

kjdchapman commented on app/views/personal_details/_form.html.erb:
```diff
@@ -1,4 +1,4 @@
-<%= render "shared/form_errors_summary", errors: @personal_details.errors %>
+<%= render "shared/form_errors_summary", model: @personal_details %>
```
Addressed in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/112/commits/6eba2e9b2e88076625d33340191f8cee8ca8d752

duncanjbrown commented on app/helpers/error_summary_helper.rb:
```diff
@@ -0,0 +1,5 @@
+module ErrorSummaryHelper
+  def anchor_link(record, field)
```
I think `#field_anchor_link`, which is in the spec, is a better name

kjdchapman replied:
I went back and forth on whether to put "field" in the module name or method name, and accidentally settled on the _worst option_, neither. Will re-introduce it.

kjdchapman replied:
Addressed in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/112/commits/a22c48dd0de9badb194a01e7f42c5a2c96fc338f

kjdchapman commented on app/views/shared/_form_errors_summary.html.erb:
```diff
@@ -1,16 +1,18 @@
+        <ul class="govuk-list govuk-error-summary__list">
+          <% model.errors.keys.each do |field| %>
+            <li>
+              <a href='#<%= model.model_name.param_key %>_<%= field %>'>
```
Had a go with a helper in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/112/commits/286ea46f3c05498a0155c4293e4c99f03450a281.

I don't mind this implementation, but I don't love the test I've ended up with (having to chain seems bleh).

duncanjbrown commented on spec/helpers/error_summary_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+  describe '#field_anchor_link' do
+    context 'given a ActiveModel record and a field_name' do
+      let(:model_name_mock) { double(param_key: 'active_model') }
+      let(:active_model_mock) { double(model_name: model_name_mock) }
```
You say in the `context` that this is an `ActiveModel record`, but it's a duck — a mock duck! — quacking like one. Why not use the real thing? As a bonus, the spec will then break usefully if the interface changes.

kjdchapman replied:
I mocked out of habit (since this is a unit spec and that's a collaborator), but I think you're right.

After reading around it a bit ("[don't mock what you don't own](https://github.com/testdouble/contributing-tests/wiki/Don't-mock-what-you-don't-own)"), I see that I don't really get any value from mocking here.

kjdchapman replied:
Addressed in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/112/commits/98b95042bc27b1476159bae1a21493a025147be7 - keen to get feedback about how clear the inline class declaration is.

duncanjbrown replied:
I've had a chance to mull this over a bit.

1. I felt uneasy about the mocking for the same reason you weren't sure about it: the chaining. I don't think mocking is necessarily wrong here.
1. Calling `model.model_name.param_key` is a Law of Demeter violation. .I know LoD violations aren't always fixable in a nice way but it's usually worth talking about when we do break it.
1. The interface we care about in this method is the one that implements `param_key`.
1. One simplification is to pass in the thing that implements `param_key`, ie the `ActiveModel::Name` itself (https://api.rubyonrails.org/classes/ActiveModel/Name.html) instead of passing the model itself. That makes this method dumber, and the call becomes more explicit.

The view code would then be

```
<a href='<%= field_anchor_link(record.model_name, field) %>'>
```

Or even with kwargs, which I generally favour:

```
<a href='<%= field_anchor_link(model_name: record.model_name, field_name: field) %>'>
```

The spec could then indeed use a mock and wouldn't read so oddly.

What do you think?

kjdchapman replied:
Had a conversation irl about this, then (hopefully!) resolved it in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/112/commits/d622cdd1379e54d3f7b31b438bc6a86c013249bd.

kjdchapman replied:
Oh, excuse me, I missed the comment on kwargs, just fixing that...

kjdchapman commented:
I think I've addressed all your feedback @duncanjbrown - we don't really have an established etiquette yet for who clicks "Resolve conversation", so I'll leave everything open for the mo.

I intend to rebase this branch before it's merged, but keeping it separate helps display the feedback changes for now.
